File: 003/Erlang/problem003.erl
-module(problem003). 
-export([greatest_prime_factor/1]).

greatest_prime_factor(N) -> gpfSearch(N, 2).

gpfSearch(N, F) ->
  if
    N == F       -> N;
    N rem F == 0 -> gpfSearch(N div F, F);
    true         -> gpfSearch(N, F + 1)
  end.

File: 003/Ruby/problem003.rb
class Integer
  def greatest_prime_factor
    self.gpfSearch(2)
  end
  
  def gpfSearch(f)
    n = self
    while n != f
      if n % f == 0
        n = n / f
      else
        f = f + 1
      end
    end
    n
  end
end
